java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java

```diff
@@ -1245,7 +1245,7 @@ protected final void handleSafe(int index, int dataLength) {
     while (index >= getValueCapacity()) {
       reallocValidityAndOffsetBuffers();
     }
-    final int startOffset = getStartOffset(index);
+    final int startOffset = lastSet < 0 ? 0 : getStartOffset(lastSet + 1);
```

Comment:
@eyalfa Thank you for taking time to fix this problem. 

I think this solution solves your problem. However, maybe it does not solve the problem completely. 

The fundamental problem is that the offset buffer is in an inconsistent state. This causes the `handleSafe` method to fail. 

The solution here tries to work around it, and the failure goes away, but the problem remains unsolved. To see this, please note that when returning to the `copySafe` method, we update the `lastSet` member without fixing the inconsistent state of the offset buffer, so the offset values in the range [lastSet, thisIndex) remains inconsistent, and we no longer have a chance to fix it. 

IMO, to fix the problem, we need to fix the problem of inconsistent state of the offset buffer. This can be achieved by calling `fillHoles` method. (Maybe there are other ways, but the point is that the problem must be solved, not bypassed)

Reply:
copyFromSafe calls fillHoles immediately after handleSafe, so at least this aspect of the fix holds.
I'd need soe more time to analyse all of the handleSafe call sites but the first thing I've noticed is that it's protected hence its usage is limited to library code, more specifically library code which is aware of the safety aspects of handling variable width vectors.
even if the handleSafe method called fillHoles there will still be potential use-cases where the offsets array can be compromised beyond repair, i.e. code that explicitly calls allocateNew, then some unsafe operations followed by 'safe' operations - the safe operation may fill the last hole but not one preceding the last set offset.

having that said, I can change my fix to invoke fillHoles but I'm not sure what does it do to the method's contract as it'd start modifying the visible state of the vector.

Reply:
> copyFromSafe calls fillHoles immediately after handleSafe, so at least this aspect of the fix holds.

You are right. Method fillHoles is called in copyFromSafe (sorry I ignored this at first). So I think your solution solves the problem for this scenario.
 
> I'd need soe more time to analyse all of the handleSafe call sites but the first thing I've noticed is that it's protected hence its usage is limited to library code, more specifically library code which is aware of the safety aspects of handling variable width vectors.
even if the handleSafe method called fillHoles there will still be potential use-cases where the offsets array can be compromised beyond repair, i.e. code that explicitly calls allocateNew, then some unsafe operations followed by 'safe' operations - the safe operation may fill the last hole but not one preceding the last set offset.

Please go ahead with your analysis, and please note that we cannot make sure the offset buffer is in a consistent state for all scenarios. We should try our best to do so, but if we cannot, we only need to support vectors which go through the life cycle in the right order (https://github.com/apache/arrow/blob/master/docs/source/java/vector.rst).

> having that said, I can change my fix to invoke fillHoles but I'm not sure what does it do to the method's contract as it'd start modifying the visible state of the vector.

I think there is no need to change it until we reach a consensus in the discussion.

Reply:
Small nit: since `lastSet ` can only be -1, the tertiary operation is not necessary.

Reply:
> Small nit: since `lastSet ` can only be -1, the tertiary operation is not necessary.

not sure I understand why? if is't still at -1 I'm not allowed to read the length at offset 0.

Reply:
when lastSet == -1, getStartOffset(lastSet + 1) amounts to getStartOffset(0), which is legal

Reply:
> > copyFromSafe calls fillHoles immediately after handleSafe, so at least this aspect of the fix holds.
> 
> You are right. Method fillHoles is called in copyFromSafe (sorry I ignored this at first). So I think your solution solves the problem for this scenario.
> 
> > I'd need soe more time to analyse all of the handleSafe call sites but the first thing I've noticed is that it's protected hence its usage is limited to library code, more specifically library code which is aware of the safety aspects of handling variable width vectors.
> > even if the handleSafe method called fillHoles there will still be potential use-cases where the offsets array can be compromised beyond repair, i.e. code that explicitly calls allocateNew, then some unsafe operations followed by 'safe' operations - the safe operation may fill the last hole but not one preceding the last set offset.
> 
> Please go ahead with your analysis, and please note that we cannot make sure the offset buffer is in a consistent state for all scenarios. We should try our best to do so, but if we cannot, we only need to support vectors which go through the life cycle in the right order (https://github.com/apache/arrow/blob/master/docs/source/java/vector.rst).
> 
> > having that said, I can change my fix to invoke fillHoles but I'm not sure what does it do to the method's contract as it'd start modifying the visible state of the vector.
> 
> I think there is no need to change it until we reach a consensus in the discussion.

Took a a tour of the code (courtesy of Idea's find references).
seems like `handleSafe` is called by `fillEmpties` and various implementations of `setSafe` (across three classes).
It seems all `setSafe` overloads calls `fillEmpties` followed by `handleSafe` which means they actually invoke `handleSafe` twice, not sure if this is intentional but it effectively eliminates the bug the PR is trying to fix, paying a performance hit while doing so. I suggest leaving my fix in place, in addition remove the redundant `handleSafe` invocations (where following `fillEmpties`).

Reply:
> when lastSet == -1, getStartOffset(lastSet + 1) amounts to getStartOffset(0), which is legal

hmm, unless `index < 0` as in this case the index array may still be the empty buffer. BTW index is not asserted to be non-negative in the `handleSafe` method and in at least two of the call sites (`fillEmpties` and `copyFromSafe`).

Reply:
> It seems all setSafe overloads calls fillEmpties followed by handleSafe which means they actually invoke handleSafe twice, not sure if this is intentional but it effectively eliminates the bug the PR is trying to fix, paying a performance hit while doing so. I suggest leaving my fix in place, in addition remove the redundant handleSafe invocations (where following fillEmpties).

Thank you for the investigation. 
So the effective call sequence is handleSafe + fillHoles + handleSafe. The first handleSafe call ensures enough capacity for the offset buffer, the second fillHoles call makes the content of the offset buffer valid, and the last handleSafe call ensures enough capacity for the data buffer. So the bug is hidden. 

Anyway, the bug in handleSafe must be fixed, and by fixing it, the second handleSafe call can be removed to improve performance (as you have suggested).

Reply:
Would you please also fix the setSafe overloads?

Reply:
I definitely intend to fix the `setSafe` overloads.
please notice my fix actually addresses your concern since one call will both ensure offsets buffer and value buffer has enough capacity for the intended update.

please notice the reason we've encountered this in `copyFromSafe` is because this method calls `handleSafe` just once.

I hope I'll be able to push a commit later today addressing the setSafe call sites.

Reply:
> I definitely intend to fix the `setSafe` overloads.
> please notice my fix actually addresses your concern since one call will both ensure offsets buffer and value buffer has enough capacity for the intended update.
> 
> please notice the reason we've encountered this in `copyFromSafe` is because this method calls `handleSafe` just once.
> 
> I hope I'll be able to push a commit later today addressing the setSafe call sites.

Sounds good. 
So maybe we need to replace fillEmpties with explicit calls to handleSafe and fillHoles. 
Please go ahead with your changes.

Reply:
> > when lastSet == -1, getStartOffset(lastSet + 1) amounts to getStartOffset(0), which is legal
> 
> hmm, unless `index < 0` as in this case the index array may still be the empty buffer. BTW index is not asserted to be non-negative in the `handleSafe` method and in at least two of the call sites (`fillEmpties` and `copyFromSafe`).

IMO, this should not be the case. However, I think it is sufficient for this fix. Maybe we can discuss it in a separate issue.